File: muk-angular/src/app/services/domiciliario.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Domiciliario, DomiciliarioUpsertRequest } from '../models/domiciliario';

/**
 * Servicio para gestión de Domiciliarios
 * 
 * Responsabilidades:
 * - CRUD de domiciliarios
 * - Activar/desactivar domiciliarios
 * - Consultar domiciliarios disponibles
 */
@Injectable({
  providedIn: 'root'
})
export class DomiciliarioService {
  private readonly apiUrl = '/api/domiciliarios';

  constructor(private readonly http: HttpClient) {}

  /**
   * Obtiene todos los domiciliarios
   */
  listarTodos(): Observable<Domiciliario[]> {
    return this.http.get<Domiciliario[]>(this.apiUrl);
  }

  /**
   * Obtiene un domiciliario por su ID
   */
  obtenerPorId(id: number): Observable<Domiciliario> {
    return this.http.get<Domiciliario>(`${this.apiUrl}/${id}`);
  }

  /**
   * Obtiene todos los domiciliarios que están activos y disponibles
   * (pueden recibir nuevas asignaciones de pedidos)
   */
  obtenerActivosDisponibles(): Observable<Domiciliario[]> {
    return this.http.get<Domiciliario[]>(`${this.apiUrl}/activos/disponibles`);
  }

  /**
   * Crea un nuevo domiciliario
   */
  crear(request: DomiciliarioUpsertRequest): Observable<Domiciliario> {
    return this.http.post<Domiciliario>(this.apiUrl, request);
  }

  /**
   * Actualiza un domiciliario existente
   */
  actualizar(id: number, request: DomiciliarioUpsertRequest): Observable<Domiciliario> {
    return this.http.put<Domiciliario>(`${this.apiUrl}/${id}`, request);
  }

  /**
   * Elimina un domiciliario
   */
  eliminar(id: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.apiUrl}/${id}`);
  }

  /**
   * Activa un domiciliario (marca como activo=true)
   */
  activar(id: number): Observable<Domiciliario> {
    return this.http.put<Domiciliario>(`${this.apiUrl}/${id}/activar`, {});
  }

  /**
   * Desactiva un domiciliario (marca como activo=false)
   */
  desactivar(id: number): Observable<Domiciliario> {
    return this.http.put<Domiciliario>(`${this.apiUrl}/${id}/desactivar`, {});
  }
}

File: src/main/java/com/muk/service/DomiciliarioService.java
package com.muk.service;

import com.muk.entities.Domiciliario;

import java.util.List;
import java.util.Optional;

/**
 * Interfaz del servicio de domiciliarios. Define la lógica de negocio.
 * 
 * Responsabilidades:
 * - Gestionar domiciliarios (CRUD)
 * - Activar/desactivar domiciliarios
 * - Consultar disponibilidad
 */
public interface DomiciliarioService {

    /**
     * Obtiene todos los domiciliarios del sistema.
     */
    List<Domiciliario> findAll();

    /**
     * Obtiene un domiciliario por su ID.
     */
    Optional<Domiciliario> findById(Long id);

    /**
     * Obtiene un domiciliario por su número de celular.
     */
    Optional<Domiciliario> findByCelular(String celular);

    /**
     * Obtiene un domiciliario por su número de cédula.
     */
    Optional<Domiciliario> findByCedula(String cedula);

    /**
     * Obtiene todos los domiciliarios activos.
     */
    List<Domiciliario> findAllActivos();

    /**
     * Obtiene todos los domiciliarios que están activos Y disponibles.
     * Estos pueden recibir nuevas asignaciones de pedidos.
     */
    List<Domiciliario> findAllActivosDisponibles();

    /**
     * Crea un nuevo domiciliario.
     * Valida que no exista otro con el mismo celular o cédula.
     */
    Domiciliario crear(Domiciliario domiciliario);

    /**
     * Actualiza un domiciliario existente.
     */
    Domiciliario actualizar(Long id, Domiciliario datosActualizados);

    /**
     * Elimina un domiciliario del sistema.
     */
    void eliminar(Long id);

    /**
     * Activa un domiciliario (marca como activo=true).
     * Un domiciliario activo puede ser gestionado y asignado a pedidos.
     */
    Domiciliario activar(Long id);

    /**
     * Desactiva un domiciliario (marca como activo=false).
     * Un domiciliario inactivo no puede recibir nuevas asignaciones.
     */
    Domiciliario desactivar(Long id);

    /**
     * Marca un domiciliario como disponible (disponible=true).
     * Solo los domiciliarios disponibles pueden recibir nuevos pedidos.
     */
    Domiciliario marcarDisponible(Long id);

    /**
     * Marca un domiciliario como no disponible (disponible=false).
     * Usado cuando el domiciliario tiene un pedido en camino.
     */
    Domiciliario marcarNoDisponible(Long id);
}
